tinygrad/tensor.py

```diff
@@ -832,7 +832,7 @@ def pow(self, x:Union[Tensor, Scalar], reverse=False) -> Tensor:
     # inject nan if the base is negative and the power is not an integer
     to_nan = (((x - x.trunc()) * 1e10).abs().clip(0, 1) if isinstance(x, Tensor) else int(bool(x - int(x))) if not reverse else ((self - self.trunc()) * 1e10).abs().clip(0, 1)) * base_sign  # noqa: E501
     inject_nan = ((((-to_nan) * 2) + 1)).log().add(1) if isinstance(to_nan, Tensor) else 1 if not to_nan else float("nan")
-    return ar.mul(sign * base_sign + (1 - base_sign)).mul(inject_nan)
+    return ar.mul(sign * base_sign + (1 - base_sign)).mul(inject_nan).cast(least_upper_dtype(self.dtype, x.dtype) if isinstance(x, Tensor) else self.dtype)  # noqa: E501
```

Comment:
do you need to also cast in line 824? please double check to make sure it's complete

Reply:
Good catch. The way the logic is now casting won't work, as the used formula to calculate the reversed power under the defined conditions doesn't produce accurate results. For example, `5**1` returns 4.9999995, so casting to `int` (in this example) will cause a loss of information. Also, while testing with scalars, I noticed the current code doesn't consider scalars' data types, so I fixed the code and added a regression test.

Reply:
either adding `Tensor.round` or change how `pow` works for int can work right?

i am closing this pr. i have given many feedback and it's still not close to being merged

Reply:
Exactly, it is between supporting `Tensor.round` or fixing `Tensor.pow` logic for scalers when `Reverse=True`.

No problem, I understand. I will be trying to do either of the above and open a new PR when I have a complete solution.